src/main/java/it/gov/pagopa/ranking/service/RankingMaterializerServiceImpl.java

```diff
@@ -0,0 +1,183 @@
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+@Slf4j
+public class RankingMaterializerServiceImpl implements RankingMaterializerService {
+
+    /** Max size of the list that will be passed to the repository saveAll method */
+    private final int savableEntitiesMaxSize;
+
+    private final OnboardingRankingRequestsRepository onboardingRankingRequestsRepository;
+    private final int size;
+    private final String tmpDir;
+    private final OnboardingRankingRequests2RankingCsvDTOMapper csvMapper;
+    private final RankingCsvWriterService csvWriterService;
+
+    public RankingMaterializerServiceImpl(
+            @Value("${app.ranking.savable-entities-size}") int savableEntitiesMaxSize,
+            OnboardingRankingRequestsRepository onboardingRankingRequestsRepository,
+            @Value("${app.ranking.query-page-size}")  int size,
+            @Value("${app.ranking.csv.tmp-dir}") String tmpDir,
+            OnboardingRankingRequests2RankingCsvDTOMapper csvMapper,
+            RankingCsvWriterService csvWriterService) {
+        this.savableEntitiesMaxSize = savableEntitiesMaxSize;
+        this.onboardingRankingRequestsRepository = onboardingRankingRequestsRepository;
+        this.size = size;
+        this.tmpDir = tmpDir;
+        this.csvMapper = csvMapper;
+        this.csvWriterService = csvWriterService;
+    }
+
+    @Override
+    public Path materialize(InitiativeConfig initiativeConfig) {
+        initiativeConfig.setRankingFilePath(buildRankingFilePath(initiativeConfig).replace(tmpDir+"/",""));
+
+        String localFileName = "%s/%s".formatted(tmpDir, initiativeConfig.getRankingFilePath());
+        createDirectoryIfNotExists(localFileName);
+
+        try (FileWriter outputCsvWriter = new FileWriter(localFileName)) {
+            String initiativeId = initiativeConfig.getInitiativeId();
+            initiativeConfig.setTotalEligibleOk(0);
+            initiativeConfig.setTotalEligibleKo(0);
+            initiativeConfig.setTotalOnboardingKo(0);
+
+            int page = 0;
+            int rank = 1;
+            List<OnboardingRankingRequests> pageContent;
+            Sort sorting = getSorting(initiativeConfig);
+            while (!(pageContent = onboardingRankingRequestsRepository.findAllByInitiativeId(
+                    initiativeId,
+                    PageRequest.of(page++, size, sorting))
+            ).isEmpty()) {
+                log.info("[RANKING_MATERIALIZER] Reading page number {} of initiative with id {}", page, initiativeId);
+
+                List<OnboardingRankingRequests> requestsToWrite = new ArrayList<>();
+                List<OnboardingRankingRequests> requestsToSave = new ArrayList<>();
```

Comment:
we know the maxSize, thus we could provide this info when creating the collection,improving performance
```suggestion
                List<OnboardingRankingRequests> requestsToWrite = new ArrayList<>(savableEntitiesMaxSize);
                List<OnboardingRankingRequests> requestsToSave = new ArrayList<>(savableEntitiesMaxSize);
```